<polymer-element name="jso-fatigo" attributes="userData selectedOption selectedStudy studies">
    <template>
        <link rel="stylesheet" href="../jso-style.css">
        <style>
            :host {
                display: block;
                position: relative;
                box-sizing: border-box;
                background-color: #FFFFFF;
                padding: 30px 100px;

                z-index: 50000;
                margin: 0px auto 0px auto;

                width: 50%;
                border: 1px solid #c6d0da;
                transition: all 0.2s;
                box-shadow: 2px 6px 15px 8px rgba(0, 0, 0, 0.30);
            }

            .title {
                text-align: center;
                font-size: 25px;
                color: #445D76;
                margin-bottom: 10px;
            }

            .close {
                color: #aaa;
            }

            .close:hover {
                color: #ffad76;
            }

            label {
                display: inline-block;
                color: #666;
                margin-top: 3px;
                /*width: 150px;*/
            }

            label::after {
                content: ':'
            }

            .action {
                /*margin-top: 30px;*/
                background-color: #f5f5f5 !important;
                text-align: center;
            }

            .action:active {
                background-color: #445D76 !important;
                color: #fff !important;
            }

            .message {
                margin-top: 20px;
            }

            div.button {
                /*display: inline-block;*/
                width: 300px
            }

            .inputListTextArea {
                width: 300px;
                height: 50px;
            }

            /*jso-file-browser {*/
            /*width: 70%;*/
            /*}*/

            span {
                border: 1px solid;
            }

            .botbar {
                margin-top: 10px;
            }
        </style>
        <div class="title" horizontal layout>
            <div flex>
                Go enrichment analysis
            </div>
            <div class="close" on-click="{{handleCancel}}">
                <font-awesome icon="times"></font-awesome>
                <!--<font-awesome icon="magic"></font-awesome>-->
            </div>
        </div>
        <form id="form">
            <div class="formbox">
                <div class="formtitle">Examples</div>
                <div class="formcontent">
                    <div class="button action" on-click="{{loadExample1}}">motor vs apoptosis</div>
                </div>
            </div>
            <div class="formbox">
                <div class="formtitle">Output folder</div>
                <div class="formcontent">
                    <jso-project-browser-window id="outdir" studies="{{studies}}" selectedStudy="{{selectedStudy}}"
                                                selectionMode="folder"></jso-project-browser-window>
                </div>
            </div>

            <div class="formbox">
                <div class="formtitle">Define your comparison</div>
                <div class="formcontent">
                    <input type="radio" name="comparisonRadio" value="list2list" on-click="{{comparisonChange}}" checked>Id list vs
                    Id
                    list<br>
                    <input type="radio" name="comparisonRadio" value="list2genome" on-click="{{comparisonChange}}">Id List vs Rest
                    of
                    genome<br>
                    <input type="radio" name="comparisonRadio" value="list2rest" on-click="{{comparisonChange}}">Id List vs Rest of
                    ids
                    contained in your annotations
                    (complementary list)<br>
                </div>
            </div>

            <div class="formbox">
                <div class="formtitle">Select your data</div>
                <div class="formcontent">
                    <jso-project-browser-window id="inputFile1" studies="{{studies}}" selectedStudy="{{selectedStudy}}"
                                                selectionMode="file" on-file-select="{{handleFileSelect}}"></jso-project-browser-window>
                </div>
            </div>

            <div class="formbox">
                <div class="formtitle">Select your data(List 2)</div>
                <div class="formcontent">
                    <jso-project-browser-window id="inputFile2" studies="{{studies}}" selectedStudy="{{selectedStudy}}"
                                                selectionMode="file" on-file-select="{{handleFileSelect}}"></jso-project-browser-window>
                </div>
            </div>
            <div class="formbox">
                <div class="formtitle">List 2</div>
                <div class="formcontent">
                    <div id="option2" hidden>
                        <span>Rest of the genome</span>
                    </div>

                    <div id="option3" hidden>
                        <span>Rest of ids contained in your annotations (complementary list)</span>
                    </div>
                </div>
            </div>
            <div class="formbox">
                <div class="formtitle">Options</div>
                <div class="formcontent">
                    <div id="options">
                        <label>Fisher exact test</label>
                        <select name="fisher" id="fisher" class="fc-combobox" type="text" value="">
                            <option name="fisher_1" id="fisher_1" class="fc-combobox-item" value="two-tailed">Two tailed</option>
                            <option name="fisher_2" id="fisher_2" class="fc-combobox-item" value="greater">Over-represented terms in list 1 (genome
                                comparison)
                            </option>
                            <option name="fisher_3" id="fisher_3" class="fc-combobox-item" value="less">Over-represented terms in list 2</option>
                        </select>
                        <label>Remove duplicates</label>
                        <select name="duplicates" id="duplicates" class="fc-combobox" type="text" value="">
                            <option name="duplicates_1" id="duplicates_1" class="fc-combobox-item" value="never">Never</option>
                            <option name="duplicates_2" id="duplicates_2" class="fc-combobox-item" value="each">Remove on each list separately</option>
                            <option name="duplicates_3" id="duplicates_3" class="fc-combobox-item" value="all">Remove on each list and common ids</option>
                            <option name="duplicates_4" id="duplicates_4" class="fc-combobox-item" value="ref">Remove from list2 those appearing in list1
                                (complementary list)
                            </option>
                        </select>
                    </div>
                </div>
            </div>
            <div class="formbox">
                <div class="formtitle">Databases</div>
                <div class="formcontent">
                    <div id="databases">
                        <select name="species" id="species" class="fc-combobox" type="text" value="none">
                            <option name="species_1" id="species_1" class="fc-combobox-item" value="none" selected="selected">Select an organism</option>
                            <option name="species_2" id="species_2" class="fc-combobox-item" value="hsa">Human (homo sapiens)</option>
                            <!--<option name="species_3" id="species_3" class="fc-combobox-item" value="mmu">Mouse (mus musculus)</option>-->
                            <!--<option name="species_4" id="species_4" class="fc-combobox-item" value="rno">Rat (rattus norvegicus)</option>-->
                            <!--<option name="species_5" id="species_5" class="fc-combobox-item" value="dme">Fruitfly (drosophila melanogaster)</option>-->
                            <!--<option name="species_6" id="species_6" class="fc-combobox-item" value="bta">Cow (bos taurus)</option>-->
                            <!--<option name="species_7" id="species_7" class="fc-combobox-item" value="dre">Zebrafish (danio rerio)</option>-->
                            <!--<option name="species_8" id="species_8" class="fc-combobox-item" value="sce">Saccharomyces cerevisiae</option>-->
                            <!--<option name="species_9" id="species_9" class="fc-combobox-item" value="cel">Caenorhabditis elegans</option>-->
                            <!--<option name="species_10" id="species_10" class="fc-combobox-item" value="ath">Arabidopsis thaliana</option>-->
                        </select>

                        <input name="go_bp" id="go_bp" type="checkbox" value="gobp"> GO biological process <br>
                        <!--<input name="go_mf" id="go_mf" type="checkbox" value="gomf" disabled>GO molecular function <br>-->
                        <!--<input name="go_cc" id="go_cc" type="checkbox" value="gocc" disabled>GO cellular component <br>-->
                        <!--<input name="go_slim" id="go_slim" type="checkbox" value="go-slim" disabled>GOSlim GOA <br>-->
                        <!--<input name="interpro" id="interpro" type="checkbox" value="interpro" disabled>Interpro <br>-->
                        <!--<input name="kegg" id="kegg" type="checkbox" value="kegg" disabled>KEGG pathways <br>-->
                        <!--<input name="reactome" id="reactome" type="checkbox" value="reactome" disabled>Reactome <br>-->
                        <!--<input name="biocarta" id="biocarta" type="checkbox" value="biocarta" disabled>Biocarta <br>-->
                        <!--<input name="mirna" id="mirna" type="checkbox" value="mirna" disabled>miRNA targets <br>-->
                        <!--<input name="jaspar" id="jaspar" type="checkbox" value="jaspar" disabled>Jaspar TFBS <br>-->
                        <!--<input name="your_annotations" id="your_annotations" type="checkbox" value="true">Your annotations <br>-->

                    </div>
                </div>
            </div>
            <div class="formbox">
                <div class="formtitle">Job information</div>
                <div class="formcontent">
                    <label>Job name</label> <input type="text" id="jobName" value="{{jobName}}" required>
                    <label>Description</label>
                    <textarea value="{{description}}" required></textarea>
                </div>
            </div>

            <div class="botbar" horizontal layout center-justified>
                <div class="button action" style="width: 150px;" on-click="{{handleForm}}">
                    <font-awesome icon="rocket"></font-awesome>
                    &nbsp; Launch job
                </div>
            </div>
        </form>
        <div class="message">{{message}}</div>
    </template>
    <script>
        Polymer({
            created: function () {
                this.jobName = 'JobName';
                this.toolName = "fatigo";
                this.description= "Description";

                /** Tool input parameters **/
                this.annotations = "";	//    --annotations <arg>              Your own annotations
                this.biocarta = false;	//    --biocarta                       biocarta
                this.biocarta_max_num_genes = 1000;	//    --biocarta-max-num-genes <arg>   biocarta max number of genes filter
                this.biocarta_min_num_genes = 5;//    --biocarta-min-num-genes <arg>   biocarta min number of genes filter
                this.biocarta_nannot_domain = false;//    --biocarta-nannot-domain         biocarta computes the number of annotated genes from all genome, otherwise from you input list
                this.chromosomes = "";	//    --chromosomes <arg>              the chromosome region to compare with list #1. Use chromosome-start and chromosome-end options tp delimite the region within the chromosome
                this.duplicates = "never";	//    --duplicates <arg>               to remove duplicated IDs, valid values: never, each, ref. By default, never
                this.fisher = "two-tailed";	//    --fisher <arg>                   the Fisher test mode, valid values: less, greater, two-tailed. By default, two-tailed
                this.genome = false;	//    --genome                         compares list #1 with the rest of genome
                this.go_bp = false;	//    --go-bp                          GO biological process database
                this.go_bp_keyword_filter = false;	//    --go-bp-keyword-filter           GO biological process, enables keywords filter
                this.go_bp_keyword_list = false;	//    --go-bp-keyword-list <arg>       GO biological process, keywords filter
                this.go_bp_keyword_operator = false;	//    --go-bp-keyword-operator <arg>   GO biological process, keywords filter logic: all or any
                this.go_bp_max_level = 15;	//    --go-bp-max-level <arg>          GO biological process, max GO level to take into account, default 15
                this.go_bp_max_num_genes = 1000;	//    --go-bp-max-num-genes <arg>      GO biological process, max number of genes filter
                this.go_bp_min_level = 3;	//    --go-bp-min-level <arg>          GO biological process, min go level to take into account, default 3
                this.go_bp_min_num_genes = 5;	//    --go-bp-min-num-genes <arg>      GO biological process, min number of genes filter
                this.go_bp_nannot_domain = false;	//    --go-bp-nannot-domain <arg>      GO biological process, computes the number of annotated genes from all genome, otherwise from your input list
                this.go_bp_propagation = false;	//    --go-bp-propagation <arg>        GO biological process, direct vs propagated annotation
                this.go_cc = false;	//    --go-cc                          GO cellular component database
                this.go_cc_keyword_filter = false;	//    --go-cc-keyword-filter           GO cellular component, enables keywords filter
                this.go_cc_keyword_list = false;	//    --go-cc-keyword-list <arg>       GO cellular component, keywords filter
                this.go_cc_keyword_operator = false;	//    --go-cc-keyword-operator <arg>   GO cellular component, keywords filter logic: all or any
                this.go_cc_max_level = false;	//    --go-cc-max-level <arg>          GO cellular component, max GO level to take into account, default 15
                this.go_cc_max_num_genes = false;	//    --go-cc-max-num-genes <arg>      GO cellular component, max number of genes filter
                this.go_cc_min_level = false;	//    --go-cc-min-level <arg>          GO cellular component, min go level to take into account, default 3
                this.go_cc_min_num_genes = false;	//    --go-cc-min-num-genes <arg>      GO cellular component, min number of genes filter
                this.go_cc_nannot_domain = false;	//    --go-cc-nannot-domain <arg>      GO cellular component, computes the number of annotated genes from all genome, otherwise from your input list
                this.go_cc_propagation = false;	//    --go-cc-propagation <arg>        GO cellular component, direct vs propagated annotation
                this.go_dag = false;	//    --go-dag <arg>                   Compute DAG for significant GO terms
                this.go_mf = false;	//    --go-mf                          GO molecular function database
                this.go_mf_keyword_filter = false;	//    --go-mf-keyword-filter           GO molecular function, enables keywords filter
                this.go_mf_keyword_list = false;	//    --go-mf-keyword-list <arg>       GO molecular function, keywords filter
                this.go_mf_keyword_operator = false;	//    --go-mf-keyword-operator <arg>   GO molecular function, keywords filter logic: all or any
                this.go_mf_max_level = false;	//    --go-mf-max-level <arg>          GO molecular function, max GO level to take into account, default 15
                this.go_mf_max_num_genes = false;	//    --go-mf-max-num-genes <arg>      GO molecular function, max number of genes filter
                this.go_mf_min_level = false;	//    --go-mf-min-level <arg>          GO molecular function, min go level to take into account, default 3
                this.go_mf_min_num_genes = false;	//    --go-mf-min-num-genes <arg>      GO molecular function, min number of genes filter
                this.go_mf_nannot_domain = false;	//    --go-mf-nannot-domain <arg>      GO molecular function, computes the number of annotated genes from all genome, otherwise from your input list
                this.go_mf_propagation = false;	//    --go-mf-propagation <arg>        GO molecular function, direct vs propagated annotation
                this.go_slim = false;	//    --go-slim                        go-slim
                this.go_slim_max_num_genes = false;	//    --go-slim-max-num-genes <arg>    go-slim max number of genes filter
                this.go_slim_min_num_genes = false;	//    --go-slim-min-num-genes <arg>    go-slim min number of genes filter
                this.go_slim_nannot_domain = false;	//    --go-slim-nannot-domain          go-slim computes the number of annotated genes from all genome, otherwise from you input list
                this.interpro = false;	//    --interpro                       interpro
                this.interpro_max_num_genes = false;	//    --interpro-max-num-genes <arg>   interpro max number of genes filter
                this.interpro_min_num_genes = false;	//    --interpro-min-num-genes <arg>   interpro min number of genes filter
                this.interpro_nannot_domain = false;	//    --interpro-nannot-domain         interpro computes the number of annotated genes from all genome, otherwise from you input list
                this.jaspar = false;	//    --jaspar                         jaspar
                this.jaspar_max_num_genes = false;	//    --jaspar-max-num-genes <arg>     jaspar max number of genes filter
                this.jaspar_min_num_genes = false;	//    --jaspar-min-num-genes <arg>     jaspar min number of genes filter
                this.jaspar_nannot_domain = false;	//    --jaspar-nannot-domain           jaspar computes the number of annotated genes from all genome, otherwise from you input list
                this.kegg = false;	//    --kegg                           kegg
                this.kegg_max_num_genes = false;	//    --kegg-max-num-genes <arg>       kegg max number of genes filter
                this.kegg_min_num_genes = false;	//    --kegg-min-num-genes <arg>       kegg min number of genes filter
                this.kegg_nannot_domain = false;	//    --kegg-nannot-domain             kegg computes the number of annotated genes from all genome, otherwise from you input list
                this.list1 = false;	//    --list1 <arg>                    the feature data containig the list #1 of genes, or the feature data file
                this.list2 = false;	//    --list2 <arg>                    the file containig the list #2 of genes, or the feature data file
                this.mirna = false;	//    --mirna                          mirna
                this.mirna_max_num_genes = false;	//    --mirna-max-num-genes <arg>      mirna max number of genes filter
                this.mirna_min_num_genes = false;	//    --mirna-min-num-genes <arg>      mirna min number of genes filter
                this.mirna_nannot_domain = false;	//    --mirna-nannot-domain            mirna computes the number of annotated genes from all genome, otherwise from you input list
                this.outdir = false;	// -o,--outdir                         outdir to save the results
                this.oreganno = false;	//    --oreganno                       oreganno
                this.oreganno_max_num_genes = false;	//    --oreganno-max-num-genes <arg>   oreganno max number of genes filter
                this.oreganno_min_num_genes = false;	//    --oreganno-min-num-genes <arg>   oreganno min number of genes filter
                this.oreganno_nannot_domain = false;	//    --oreganno-nannot-domain         oreganno computes the number of annotated genes from all genome, otherwise from you input list
                this.reactome = false;	//    --reactome                       reactome
                this.reactome_max_num_genes = false;	//    --reactome-max-num-genes <arg>   reactome max number of genes filter
                this.reactome_min_num_genes = false;	//    --reactome-min-num-genes <arg>   reactome min number of genes filter
                this.reactome_nannot_domain = false;	//    --reactome-nannot-domain         reactome computes the number of annotated genes from all genome, otherwise from you input list
                this.species = false;	//    --species <arg>                  The specie of the ids


            },
            comparisonChange: function () {
                this.comparison = this.shadowRoot.querySelector('input[name="comparisonRadio"]:checked').value;
                this.list2 = false;
                this.rest = false;
                this.genome = false;
                if (this.comparison == "list2list") {
                    this.$.inputFile2.hidden = false;
                    this.$.option2.hidden = true;
                    this.$.option3.hidden = true;
                }
                if (this.comparison == "list2genome") {
                    this.genome = true;
                    this.$.inputFile2.hidden = true;
                    this.$.option2.hidden = false;
                    this.$.option3.hidden = true;
                }
                if (this.comparison == "list2rest") {
                    this.genome = true;
                    this.$.inputFile2.hidden = true;
                    this.$.option2.hidden = true;
                    this.$.option3.hidden = false;
                }

            },
            handleCancel: function () {
                this.selectedOption = '';
            },
            handleForm: function () {
                this.launchJob();
            },
            launchJob: function () {
                var me = this;
                var query = {
                    sid: Cookies("bioinfo_sid"),
                    studyId: this.selectedStudy.id,
                    toolId: this.toolName,
                    name: this.jobName,
                    description: "description",
                    outdir: this.$.outdir.selectedFile.id,
                    list1: this.$.inputFile1.selectedFile.id,
                    list2: this.$.inputFile2.selectedFile.id,
                    duplicates: this.$.duplicates.value,
                    "species": this.$.species.value,
                    fisher: this.$.fisher.value
                };
                this.comparison = this.shadowRoot.querySelector('input[name="comparisonRadio"]:checked').value;
                if (this.comparison == "list2list") {
                    query["list2"] = this.$.inputFile2.selectedFile.id;
                }
                else if (this.comparison == "list2genome") {
                    query["genome"] = "";
                }
                else if (this.comparison == "list2rest") {

                }
                if (this.$.go_bp.checked) {
                    query["go-bp"] = "";
                    query["go-bp-min-num-genes"] = this.go_bp_min_num_genes;
                    query["go-bp-max-num-genes"] = this.go_bp_max_num_genes;
                }
                OpencgaManager.jobs.create({
                    query: query,
                    request: {
                        //method: 'POST',
                        success: function (response) {
                            if (response.response[0].errorMsg === '' || response.response[0].errorMsg == null) {
                                var userId = response.response[0].result[0].id;
                                me.message = userId + ' created';
                                me.selectedOption = "jobLaunched"
                            } else {
                                me.message = response.response[0].errorMsg;
                            }
                        },
                        error: function () {
                            me.message = 'Server error, try again later.';
                        }
                    }
                })
            },

            loadExample1: function (e) {
                this.$.go_bp.disabled = false;
                this.$.go_bp.checked = true;
                this.message = '';
                this.newFolderName = "fatiresult_" + Math.floor((Math.random() * 100) + 1);
                this.duplicates = "never";
                this.go_bp_keyword_operator = "all";
                this.$.species.value = "hsa";
                this.go_bp_max_level = 9;
                this.go_bp = "gobp";
                this.go_bp_propagation = "propagate";
                this.go_bp_min_num_genes = 5;
                this.$.inputFile1.selectedFile = {id: "example_example.motor", path: "example.motor"};
                this.$.inputFile2.selectedFile = {id: "example_example.apoptosis", path: "example.apoptosis"};
                this.annotations = "none";
                this.go_bp_nannot_domain = "genome";
                this.fisher = "two-tailed";
                this.go_bp_max_num_genes = 1000;
                this.go_bp_min_level = 3;

                var exampleFolder = "fati_motor_apop";
                this.createExampleFolder(exampleFolder);

            },
            createExampleFolder: function (exampleFolder) {
                var me = this;
                OpencgaManager.files.search({
                    query: {
                        sid: Cookies('bioinfo_sid'),
                        studyId: me.$.outdir.selectedStudy.id,
                        name: exampleFolder

                    },
                    request: {
                        success: function (response) {
                            if (response.response[0].errorMsg === '' || response.response[0].errorMsg == null) {
                                me.$.outdir.selectedFile = response.response[0].result[0];
                                if (response.response[0].result.length == 0) {
                                    OpencgaManager.files.createFolder({
                                        query: {
                                            sid: Cookies('bioinfo_sid'),
                                            studyId: me.$.outdir.selectedStudy.id,
                                            folder: exampleFolder

                                        },
                                        request: {
                                            success: function (response) {
                                                if (response.response[0].errorMsg === '' || response.response[0].errorMsg == null) {
                                                    me.$.outdir.selectedFile = response.response[0].result[0];
                                                } else {

                                                }
                                            },
                                            error: function () {
                                                alert('Server error, try again later.');
                                            }
                                        }
                                    });
                                }
                            }
                        },
                        error: function () {
                            alert('Server error, try again later.');
                        }
                    }
                });
            },

            handleSpecies: function (e) {
                this.species = e.target.dataset.value;
                this.speciesText = e.target.innerText;
                if (this.species == "hsa") {
                    this.$.go_bp.disabled = false;
                    this.$.go_bp.disabled = false;
                    this.$.go_mf.disabled = false;
                    this.$.go_cc.disabled = false;
                    this.$.go_slim.disabled = false;
                    this.$.interpro.disabled = false;
                    this.$.kegg.disabled = false;
                    this.$.reactome.disabled = false;
                    this.$.biocarta.disabled = false;
                    this.$.mirna.disabled = false;
                    this.$.jaspar.disabled = false;
                }

                if (this.species == "mmu" || this.species == "rno") {
                    this.$.go_bp.disabled = false;
                    this.$.go_bp.disabled = false;
                    this.$.go_mf.disabled = false;
                    this.$.go_cc.disabled = false;
                    this.$.go_slim.disabled = true;
                    this.$.interpro.disabled = false;
                    this.$.kegg.disabled = false;
                    this.$.reactome.disabled = true;
                    this.$.biocarta.disabled = true;
                    this.$.mirna.disabled = false;
                    this.$.jaspar.disabled = false;
                }
                if (this.species == "dme" || this.species == "bta" || this.species == "dre" || this.speces == "sce" || this.species == "cel" || this.species == "ath") {
                    this.$.go_bp.disabled = false;
                    this.$.go_bp.disabled = false;
                    this.$.go_mf.disabled = false;
                    this.$.go_cc.disabled = false;
                    this.$.go_slim.disabled = true;
                    this.$.interpro.disabled = false;
                    this.$.kegg.disabled = false;
                    this.$.reactome.disabled = true;
                    this.$.biocarta.disabled = true;
                    this.$.mirna.disabled = true;
                    this.$.jaspar.disabled = true;
                }

            }
        });
    </script>
</polymer-element>
