<!--
  ~ Copyright 2015 OpenCB
  ~
  ~ Licensed under the Apache License, Version 2.0 (the "License");
  ~ you may not use this file except in compliance with the License.
  ~ You may obtain a copy of the License at
  ~
  ~     http://www.apache.org/licenses/LICENSE-2.0
  ~
  ~ Unless required by applicable law or agreed to in writing, software
  ~ distributed under the License is distributed on an "AS IS" BASIS,
  ~ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
  ~ See the License for the specific language governing permissions and
  ~ limitations under the License.
  -->

<link rel="import" href="variant-browser-grid.html">
<link rel="import" href="../opencga-active-filters.html">
<link rel="import" href="../../cellbase/variation/cellbase-variantannotation-view.html">
<link rel="import" href="variant-beacon-network.html">
<link rel="import" href="../../genome-browser/genome-browser-view.html">

<dom-module id="variant-browser-view">
    <template>
        <style is="custom-style" include="jso-styles"></style>
        <style>
            .variant-tab-title {
                font-size: 115%;
                font-weight: bold;
            }

            th {
                text-align: center
            }
        </style>

        <div>
            <br>
            <opencga-active-filters opencga-client="{{opencgaClient}}" query="{{query}}" filters="{{config.filters}}" filter-bioformat="VARIANT" fixed-filters="{{fixedFilters}}"
                                    alias="{{activeFilterAlias}}" refresh="{{search}}"></opencga-active-filters>

            <ul class="nav nav-pills" style="padding-bottom: 15px">
                <li role="presentation" class="active"><a href="#{{prefix}}VariantView" role="tab" data-toggle="tab">Variant View</a></li>
                <!--<li role="presentation"><a href="#{{prefix}}GeneView" role="tab" data-toggle="tab">Gene View</a></li>-->
                <li role="presentation"><a href="#{{prefix}}GenomeBrowser" role="tab" data-toggle="tab">Genome Browser</a></li>
                <li role="presentation"><a href="#{{prefix}}Summary" role="tab" data-toggle="tab" on-click="fetchFacets">Summary</a></li>
            </ul>

            <div class="tab-content">
                <div role="tabpanel" class="tab-pane active" id="{{prefix}}VariantView">
                    <!-- Variant Browser Grid -->
                    <variant-browser-grid study="{{study}}" project="{{project}}" cohorts="{{cohorts}}" opencga-client="{{opencgaClient}}" cellbase-client="{{cellbaseClient}}"
                                          population-frequencies="{{populationFrequencies}}" protein-substitution-scores="{{proteinSubstitutionScores}}"
                                          consequence-types="{{consequenceTypes}}" query="{{query}}" search="{{search}}" variant="{{variant}}" on-selected="selectedGene"
                                          on-selectvariant="onSelectVariant" style="font-size: 12px" config="{{config}}" summary="{{summary}}">
                    </variant-browser-grid>

                    <div hidden$="{{checkVariant(variant)}}">
                        <br>
                        <h3>Please select a variant to view variant's detailed annotation</h3>
                    </div>

                    <!-- Bottom tabs with specific variant information -->
                    <div style="padding-top: 20px" hidden$="{{!checkVariant(variant)}}">
                        <h3>Variant: {{variant}}</h3>
                        <div style="padding-top: 20px">
                            <ul id="{{prefix}}ViewTabs" class="nav nav-tabs" role="tablist">
                                <li role="presentation" class="active"><a href="#{{prefix}}Annotation" role="tab" data-toggle="tab" class="variant-tab-title">Advanced Annotation</a></li>
                                <li role="presentation"><a href="#{{prefix}}Genotype" role="tab" data-toggle="tab" class="variant-tab-title">Genotype Stats</a></li>
                                <li role="presentation"><a href="#{{prefix}}GenomeView" role="tab" data-toggle="tab" class="variant-tab-title">Genome Context</a></li>
                                <li role="presentation"><a href="#{{prefix}}Beacon" role="tab" data-toggle="tab" class="variant-tab-title">Beacon Network</a></li>
                            </ul>

                            <div class="tab-content" style="height: 680px">

                                <!--Annotation Tab-->
                                <div role="tabpanel" class="tab-pane active" id="{{prefix}}Annotation">
                                    <cellbase-variantannotation-view data="{{variant}}" prefix="{{prefix}}" cellbase-client="{{cellbaseClient}}" mode="vertical"
                                                                     hash-fragment-credentials="{{hashFragmentCredentials}}" consequence-types="{{consequenceTypes}}"
                                                                     protein-substitution-scores="{{proteinSubstitutionScores}}" style="font-size: 12px"></cellbase-variantannotation-view>
                                </div>

                                <!--Genotypes & Files Tab-->
                                <div role="tabpanel" class="tab-pane" id="{{prefix}}Genotype">
                                    <template is="dom-repeat" items="{{variantStudies}}" as="study" id="variantStudies">
                                        <div class="col-md-10" style="padding-top: 8px">
                                            <h4 style="font-weight: bold;padding: 20px 0px 0px 0px">
                                                <i class="fa fa-minus-circle" on-click="handleCollapseAction" data-id="{{getStudy(study.studyId)}}" style="cursor: pointer;"></i>&nbsp;{{getStudy(study.studyId)}}
                                            </h4>

                                            <div id="{{getStudy(study.studyId)}}">
                                                <div hidden$="{{!checkStatsAvailable(study.stats)}}">
                                                    <table class="table table-bordered" style="width: 70%">
                                                        <thead style="background-color: #eee;">
                                                        <tr>
                                                            <th scope="col" rowspan="2">Cohort</th>
                                                            <th scope="col" rowspan="2">Reference</th>
                                                            <th scope="col" rowspan="2">Alternate</th>
                                                            <th scope="col" rowspan="2">MAF (allele)</th>
                                                            <th colspan="2" scope="colgroup">Allele Frequencies</th>
                                                            <th colspan="3" scope="colgroup">Genotype Frequencies</th>
                                                        </tr>
                                                        <tr>
                                                            <th scope="col">Reference (count)</th>
                                                            <th scope="col">Alternate (count)</th>
                                                            <th scope="col">0/0 (count)</th>
                                                            <th scope="col">0/1 (count)</th>
                                                            <th scope="col">1/1 (count)</th>
                                                        </tr>
                                                        </thead>
                                                        <tbody>
                                                        <template is="dom-repeat" items="{{study.stats}}" as="cohort">
                                                            <tr>
                                                                <td>{{cohort.name}}</td>
                                                                <td>{{cohort.value.refAllele}}</td>
                                                                <td>{{cohort.value.altAllele}}</td>
                                                                <td>{{cohort.value.maf}} ({{cohort.value.mafAllele}})</td>
                                                                <td>{{cohort.value.refAlleleFreq}} ({{cohort.value.refAlleleCount}})</td>
                                                                <td>{{cohort.value.altAlleleFreq}} ({{cohort.value.altAlleleCount}})</td>
                                                                <td>{{cohort.value.genotypesFreq.homref}} ({{cohort.value.genotypesCount.homref}})</td>
                                                                <td>{{cohort.value.genotypesFreq.het}} ({{cohort.value.genotypesCount.het}})</td>
                                                                <td>{{cohort.value.genotypesFreq.homalt}} ({{cohort.value.genotypesCount.homalt}})</td>
                                                            </tr>
                                                        </template>
                                                        </tbody>
                                                    </table>
                                                </div>
                                                <div hidden$="{{checkStatsAvailable(study.stats)}}">
                                                    No Genotypes statistics available
                                                </div>
                                            </div>
                                            <br>
                                        </div>
                                    </template>
                                </div>

                                <!--Genome Context Tab-->
                                <div role="tabpanel" class="tab-pane" id="{{prefix}}GenomeView">
                                    <br>
                                    <genome-browser-view cellbase-client="{{cellbaseClient}}" opencga-client="{{opencgaClient}}" region="{{region}}"></genome-browser-view>
                                    <!--This will be available in the coming days, sorry for any inconvenience.-->
                                    <!--<genome-context></genome-context>-->
                                </div>

                                <!--Beacon network-->
                                <div role="tabpanel" class="tab-pane" id="{{prefix}}Beacon">
                                    <br>
                                    <button class="btn btn-primary" type="button" on-click="triggerBeacon">Search Beacon Network</button>
                                    <a data-toggle="tooltip" title="Beacon Network is a search engine across the world's public beacons. You can find it here - https://beacon-network.org/#/"><i class="fa fa-info-circle" aria-hidden="true"></i></a>
                                    <br><br>
                                    <variant-beacon-network hosts="{{beaconHosts}}" clear="{{variant}}" variant="{{variantToBeacon}}"></variant-beacon-network>
                                </div>
                            </div>
                        </div>
                    </div>
                </div>
                <!--<div role="tabpanel" class="tab-pane" id="{{prefix}}GeneView">Yet to implement!</div>-->
                <div role="tabpanel" class="tab-pane" id="{{prefix}}GenomeBrowser">
                    <genome-browser-view cellbase-client="{{cellbaseClient}}" opencga-client="{{opencgaClient}}" region="{{region}}"></genome-browser-view>
                </div>
                <div role="tabpanel" class="tab-pane" id="{{prefix}}Summary">
                    <div class="col-md-6">
                        <h4>Consequence Type</h4>
                        <div id="{{prefix}}soAcc"></div>
                    </div>
                    <div class="col-md-6">
                        <h4>Gene</h4>
                        <div id="{{prefix}}genes"></div>
                    </div>
                    <div class="col-md-6">
                        <h4>Variant Type</h4>
                        <div id="{{prefix}}type"></div>
                    </div>
                    <div class="col-md-6">
                        <h4>Studies</h4>
                        <div id="{{prefix}}studies"></div>
                    </div>
                </div>
            </div>
        </div>
    </template>
    <script>
        Polymer({
            is: 'variant-browser-view',
            properties: {
                title: {
                    type: String
                },
                project: {
                    type: Object
                },
                study: {
                    type: Object
                },
                cohorts: {
                    type: Object
                },
                data: {
                    type: Array,
                    value: []
                },
                opencgaClient: {
                    type: Object
                },
                cellbaseClient: {
                    type: Object
                },
                populationFrequencies: {
                    type: Object
                },
                proteinSubstitutionScores: {
                    type: Object
                },
                consequenceTypes: {
                    type: Object
                },
                beaconHosts: {
                    type: Array
                },
                query: {
                    type: Object,
                    notify: true,
                },
                search: {
                    type: Object,
                },
                variant: {
                    type: String,
                    value: "No variant selected",
                    observer: 'variantChanged'
                },
                variantStudies: {
                    type: Array,
                    value: [],
//                    observer: "variantStudiesChanged"
                },
                prefix: {
                    type: String
                },
                summary: {
                    type: Boolean
                },
                config: {
                    type: Object
                }
            },
            observers: [
                'projectStudyObtained(project, study)'
            ],
            ready: function () {
                if (typeof this.prefix === "undefined" || this.prefix === "") {
                    this.prefix = "view" + Utils.randomString(6);
                }

                this.genotypeColor = {
                    "0/0": "#6698FF",
                    "0/1": "#FFA500",
                    "1/0": "#FFA500",
                    "1/1": "#FF0000",
                    "./.": "#000000",
                    "0|0": "#6698FF",
                    "0|1": "#FFA500",
                    "1|0": "#FFA500",
                    "1|1": "#FF0000",
                    ".|.": "#000000",
                };

                this.activeFilterAlias = {
                    "annot-xref": "XRef",
                    "annot-biotype": "Biotype",
                    "annot-ct": "Consequence Types",
                    "annot-population-maf": "Population Frequency",
                    "annot-functional-score": "CADD",
                    "protein_substitution": "Protein Substitution",
                    "annot-go": "GO",
                    "annot-hpo": "HPO"
                };

                this.fixedFilters = ["studies"];
            },
            projectStudyObtained: function (project, study) {
                if (typeof this.project !== "undefined" && this.project.alias !== "" && typeof this.study !== "undefined" && this.study.alias !== "") {
                    this.hashFragmentCredentials = {
                        project: this.project.alias,
                        study: this.study.alias
                    };
                }
            },
            checkVariant: function(variant) {
                return variant.split(':').length > 2;
            },
            onSelectVariant: function(e) {
                this.variant = e.detail.id;
//                this.V = e.detail.variant;
//                this.display3D();
            },
            selectedGene: function (e) {
                this.fire('propagate', {gene: e.detail.gene});
            },
            checkStatsAvailable: function (stats) {
                return stats.length > 0;
            },
            variantChanged: function (e) {
                if (this.variant.split(':').length > 2) {
                    let [chromosome, start, ref, alt] = this.variant.split(":");
                    this.region = new Region(chromosome + ":" + start);
                    let params = {
                        ids: this.variant,
                        studies: this.project.alias + ":" + this.study.alias,
                        exclude: "annotation,studies.files,studies.samplesData"
                    };
                    let _this = this;
                    this.opencgaClient.variants().query(params)
                        .then(function (response) {
                            if (typeof response.response[0].result[0] !== "undefined") {
                                let _variantStudies = response.response[0].result[0].studies;
                                for (let i = 0; i < _variantStudies.length; i++) {
                                    let study = _variantStudies[i].studyId.split(':')[1];
                                    let statsObject = _variantStudies[i].stats;
                                    let statsArray = [];
                                    Object.keys(statsObject).map(key => {
                                        statsObject[key].maf = _this._freqFormatter(statsObject[key].maf || 0);
                                        statsObject[key].mafAllele = statsObject[key].mafAllele || '-';

                                        statsObject[key].refAlleleFreq = _this._freqFormatter(statsObject[key].refAlleleFreq || 0);
                                        statsObject[key].altAlleleFreq = _this._freqFormatter(statsObject[key].altAlleleFreq || 0);

                                        statsObject[key].genotypesFreq.homref = _this._freqFormatter(statsObject[key].genotypesFreq["0/0"] || statsObject[key].genotypesFreq["0|0"] || 0);
                                        statsObject[key].genotypesFreq.het = _this._freqFormatter(statsObject[key].genotypesFreq["0/1"] || statsObject[key].genotypesFreq["0|1"] || 0);
                                        statsObject[key].genotypesFreq.homalt = _this._freqFormatter(statsObject[key].genotypesFreq["1/1"] || statsObject[key].genotypesFreq["1|1"] || 0);

                                        statsObject[key].genotypesCount.homref = statsObject[key].genotypesCount["0/0"] || statsObject[key].genotypesCount["0|0"] || 0;
                                        statsObject[key].genotypesCount.het = statsObject[key].genotypesCount["0/1"] || statsObject[key].genotypesCount["0|1"] || 0;
                                        statsObject[key].genotypesCount.homalt = statsObject[key].genotypesCount["1/1"] || statsObject[key].genotypesCount["1|1"] || 0;


                                        if (key === "ALL") {
                                            statsArray.unshift({
                                                name: key,
                                                value: statsObject[key],
                                                study: study
                                            })
                                        } else {
                                            statsArray.push({
                                                name: key,
                                                value: statsObject[key],
                                                study: study
                                            })
                                        }
                                    });
                                    _variantStudies[i].stats = statsArray;
                                }
                                _this.variantStudies = _variantStudies;
                            }
                        });
                }
            },
            fetchFacets: function () {
                let queryParams = {
                    sid: this.opencgaClient._config.sessionId,
                    studies: this.project.alias + ":" + this.study.alias,
                    facet: "type;genes;soAcc;studies",
                    timeout: 60000
                };
                Object.assign(queryParams, this.query);

                if (typeof queryParams.sid === "undefined" || queryParams.sid === "") {
                    delete queryParams["sid"];
                }

                let _this = this;

                this.opencgaClient.variants().facet(queryParams, {})
                    .then(function (queryResponse) {
                        _this.facetResults = queryResponse.response[0].result[0].result.fields;
                        for (let result of _this.facetResults) {
                            let categories = [];
                            let data =  [];

                            for (let countObj of result.counts) {
                                categories.push(countObj.value);
                                data.push(countObj.count);
                            }

                            $("#" + _this.prefix + result.name).highcharts({
                                credits: {enabled: false},
                                chart: {
                                    type: "column"
                                },
                                title: {
                                    text: result.name
                                },
                                xAxis: {
                                    categories: categories
                                },
                                yAxis: {
                                    min: 0,
                                    title: {
                                        text: 'Total number of Variants'
                                    }
                                },
                                tooltip: {
                                    headerFormat: '<span style="font-size:10px">{point.key}</span><table>',
                                    pointFormat: '<tr><td style="color:{series.color};padding:0">{series.name}: </td>' +
                                    '<td style="padding:0"><b>{point.y:.1f} </b></td></tr>',
                                    footerFormat: '</table>',
                                    shared: true,
                                    useHTML: true
                                },
                                plotOptions: {
                                    column: {
                                        pointPadding: 0.2,
                                        borderWidth: 0
                                    }
                                },
                                series: [{
                                    name: result.name,
                                    data: data
                                }]
                            });
                        }
                    });
            },
            _freqFormatter: function(value) {
                if (value !== 0 && value !== 1) {
                    return Number(value).toFixed(5);
                }
                return value;
            },
            // This observer is needed when pie charts are rendered instead of table. This is where the data needed for high charts is prepared
            variantStudiesChanged: function () {
                this.$.variantStudies.render();
                if (this.variantStudies.length > 0) {
                    for (let i = 0; i < this.variantStudies.length; i++) {
                        let statsArray = this.variantStudies[i].stats;
                        for (let j = 0; j < statsArray.length; j++) {
                            let genotypeCount = statsArray[j].value.genotypesCount;
                            let data = [];
                            if (Object.keys(genotypeCount).length > 0) {
                                for (let k in genotypeCount) {
                                    data.push({
                                        name: k,
                                        y: genotypeCount[k],
                                        color: this.genotypeColor[k]
                                    });
                                }
                                let params = {title: statsArray[j].name, data: data};
                                this._updatePieChart("#" + statsArray[j].study + statsArray[j].name, params);
                            }
                        }
                    }
                }
            },
            getStudy: function (study) {
                let name = study.split(':')[1];
                return name;
            },
            checkForGenotypes: function (stat) {
                let genotypesCount = stat.value.genotypesCount;
                return Object.keys(genotypesCount).length > 0;
            },
            checkForSampleData: function (samplesData) {
                return samplesData.length > 0;
            },
            handleCollapseAction: function (e) {
                let id = e.target.dataId;
                let elem = $('#' + id)[0];
                elem.hidden = !elem.hidden;
                if (elem.hidden) {
                    e.target.className = "fa fa-plus-circle";
                } else {
                    e.target.className = "fa fa-minus-circle";
                }
            },
            // This is the old Pie Chart implementation, not used right now
            _updatePieChart: function(_id, params) {
                // params = { title, data}
                $(_id).highcharts({
                    credits: {enabled: false},
                    chart: {
                        plotBackgroundColor: null,
                        plotBorderWidth: null,
                        plotShadow: false,
                        type: 'pie',
                        width: 220
                    },
                    title: {
                        text: params.title
                    },
                    tooltip: {
                        pointFormat: '{series.name}: <b>{point.percentage:.2f}%</b>'
                    },
                    plotOptions: {
                        pie: {
                            allowPointSelect: true,
                            cursor: 'pointer',
                            minSize: 75,
                            dataLabels: {
                                padding: 4,
                                connectorPadding: 4,
                                enabled: true,
                                format: '{point.name}: {point.percentage:.2f}%',
                                style: {
                                    color: (Highcharts.theme && Highcharts.theme.contrastTextColor) || 'black'
                                }
                            }
                        }
                    },
                    series: [{
                        name: 'Genotype Count',
                        data: params.data
                    }]
                });
            },
            triggerBeacon: function (e) {
                this.variantToBeacon = this.variant;
            }
        });
    </script>
</dom-module>
